src/network.js

```diff
@@ -67,13 +60,12 @@ class Network {
     pull(
       conn,
       lp.decode(),
-      pull.asyncMap((data, cb) => Message.deserialize(data, cb)),
+      pull.asyncMap((data, cb) => Message.deserialize(data).then(data => cb(null, data), cb)),
```

Comment:
call `cb` inside setImmediate when mixing promise and callbacks

Reply:
Why? As I understand it the function passed to `.then` is already going to be deferred to the microtask queue at end of the event loop, `setImmediate` will execute it on the next loop.

Reply:
i have answered this question many times so this time a made an example i can point to. 
https://codesandbox.io/s/mixing-promises-with-callbacks-i1w0z

hope this explains the problem.

Reply:
Gotcha.  Have updated.  We'll pull all that out once libp2p ships it's pull-streams --> async/interator refactor anyway.